### PR TITLE
Upstream: unregister dai

### PR DESCRIPTION
**Pull request by ranj063**

### CODE REVIEW

File: sound/soc/soc-core.c [hunk 1]
```diff
@@ -2874,6 +2874,29 @@ static inline char *fmt_multiple_name(struct device *dev,
...
+	struct snd_soc_dai *dai, *_dai;
+
+	for_each_component_dais_safe(component, dai, _dai) {
+		if (dai->driver == dai_drv) {
+			dev_dbg(component->dev, "ASoC: Unregistered DAI '%s'\n",
+				dai->name);
+			list_del(&dai->list);
+			kfree(dai->name);
+			kfree(dai);
+			return;
```
Comment (plbossart) [inline sound/soc/soc-core.c:2895]: this looks similar to snd_soc_unregister_dais(), except that you return here. would there be any merits in letting the loop continue?
Also the core is now quite confusing with (un)register_dais() and (un)register_dai() that are serve different purposes.

Comment (ranj063) [reply sound/soc/soc-core.c:2895]: @plbossart the main difference between unregister_dai() and unregister_dais() is the check to see if the dai->drv matches the input argument dai_drv. I could let the loop run, but I dont think there will be more than one dai with the same dai_drv.

**Comment by wenqingfu** [comment]
Can you please link it back to which issue to fix? Preferred method is to use [the GitHub way](https://blog.github.com/2013-05-14-closing-issues-via-pull-requests/) @ranj063

**Comment by plbossart** [comment]
On 1/14/19 5:26 PM, Ranjani Sridharan wrote:
>
> @plbossart <https://github.com/plbossart> the main difference between 
> unregister_dai() and unregister_dais() is the check to see if the 
> dai->drv matches the input argument dai_drv. I could let the loop run, 
> but I dont think there will be more than one dai with the same dai_drv.
>
what if you have multiple DAIs of the same type, e.g. HDMI or SSP?

**Comment by ranj063** [comment]
> Can you please link it back to which issue to fix? Preferred method is to use [the GitHub way](https://blog.github.com/2013-05-14-closing-issues-via-pull-requests/) @ranj063

@wenqingfu i have referenced the issue this fixes #530 but since this issue is a set of patches to verify, Im not sure we can close it with just one fix.

**Comment by ranj063** [comment]
> On 1/14/19 5:26 PM, Ranjani Sridharan wrote: @plbossart <https://github.com/plbossart> the main difference between unregister_dai() and unregister_dais() is the check to see if the dai->drv matches the input argument dai_drv. I could let the loop run, but I dont think there will be more than one dai with the same dai_drv.
> what if you have multiple DAIs of the same type, e.g. HDMI or SSP?

@plbossart these are the FE DAI's based on the PCM elems in topology. For each PCM elem in topology, there's a snd_soc_dai_driver and a snd_soc_dai created

**Comment by plbossart** [comment]
On 1/14/19 6:47 PM, Ranjani Sridharan wrote:
>
>     On 1/14/19 5:26 PM, Ranjani Sridharan wrote: @plbossart
>     <https://github.com/plbossart> https://github.com/plbossart the
>     main difference between unregister_dai() and unregister_dais() is
>     the check to see if the dai->drv matches the input argument
>     dai_drv. I could let the loop run, but I dont think there will be
>     more than one dai with the same dai_drv.
>     what if you have multiple DAIs of the same type, e.g. HDMI or SSP?
>
> @plbossart <https://github.com/plbossart> these are the FE DAI's based 
> on the PCM elems in topology. For each PCM elem in topology, there's a 
> snd_soc_dai_driver and a snd_soc_dai created
>
Same line of thought, can't we have the same driver handling multiple 
PCMs? Why would there be an exact match between driver and PCM?

**Comment by ranj063** [comment]
@plbossart im not the expert with dai and dai_drv. I think I need @keyonjie or @bardliao to back me up here.

But as far as I can tell irrespective of whether it is a FE dai or a BE DAI, we seem have separate dai_drv's for it. For the BE dai's, these are defined by struct snd_soc_dai_driver skl_dai[] for APL for example.

**Comment by bardliao** [comment]
@plbossart @ranj063 If we look at [snd_soc_register_dais](https://elixir.bootlin.com/linux/latest/source/sound/soc/soc-core.c#L2950) I think dai and dai_drv is 1-1 mapping in current usage. And yes, we can use a single dai_drv for same type dais, but I didn't see a strong reason to do it.

**Comment by ranj063** [comment]
closing this as #542 handles this too

**Pull request closed without merging**